datalad/auto.py

```diff
@@ -0,0 +1,184 @@
+# emacs: -*- mode: python; py-indent-offset: 4; tab-width: 4; indent-tabs-mode: nil -*-
+# ex: set sts=4 ts=4 sw=4 noet:
+# ## ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ##
+#
+#   See COPYING file distributed along with the datalad package for the
+#   copyright and license terms.
+#
+# ## ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ##
+"""Proxy basic file operations (such as open) to obtain files automagically upon I/O
+"""
+
+from mock import patch
+from six import PY2
+import six.moves.builtins as __builtin__
+builtins_name = '__builtin__' if PY2 else 'builtins'
+
+try:
+    import h5py
+except ImportError:
+    h5py = None
+
+import logging
+
+from os.path import dirname, abspath, pardir, join as opj, exists, basename
+
+from .support.annexrepo import AnnexRepo
+from .support.gitrepo import GitRepo
+from .support.exceptions import CommandError
+from .cmd import Runner
+
+from .utils import swallow_outputs
+lgr = logging.getLogger("datalad.auto")
+
+
+# TODO: shouldn't be some classmethod of GitRepo?
```

Comment:
my timing on smaug in a repo with >10,000 files:
```shell
smaug:/mnt/scrap/datalad/mimic_repos/hcp-ls.20141020-500subjects/144226/hcp-openaccess/500subject/144226/MNINonLinear
$> time git ls-files wmparc.nii.gz --error-unmatch
wmparc.nii.gz
git ls-files wmparc.nii.gz --error-unmatch  0.01s user 0.00s system 95% cpu 0.013 total
```
try to beat that one? ;) ok:
```
smaug:/mnt/scrap/datalad/mimic_repos/hcp-ls.20141020-500subjects/144226
$> python -c "import time; from datalad.support.annexrepo import GitRepo; annex=GitRepo('$PWD'); t1=time.time(); annex.get_indexed_files(); print(time.time() - t1)"
0.0757648944855
```
BUT you are correct that there might be (more) appropriate function within GitRepo. The problem though is that for that one you need to first figure out top directory which has .git, whenever ls-files  can work from wherever

re read-only -- lets' continue in #204?